### PR TITLE
change pre_autograd to pre_dispatch tracing

### PR DESCRIPTION
**Pull request by bdhirsh**
We discussed in a composability meeting a few weeks ago that `pre_autograd` should probably be renamed to `pre_dispatch`.

One question in this PR was: should I re-use a dispatch key? Or should I create a new dispatch key (that yet again corresponds to "top of the dispatcher")?

~~For now, I ended up sticking our proxy mode on the mode stack corresponding to `PythonTLSSnapshot`, because it was simple and it works. It looks like one of the functorch dispatch keys has higher priority though, so it's possible that functorch will end up running first. Open to options, but we can consider adding a new dispatch key later if that becomes a problem~~

Update: I added a dedicated dispatch key, `PreDispatch`.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #103026
* #103025
* #103024
* __->__ #101818

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @aakhundov @desertfire

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

**Review by Neilblaze — APPROVED** [review]

**Review by albanD — APPROVED** [review]

**Comment by bdhirsh** [comment]
The major update to this PR is that:

- I added a dedicated `PreDispatch` dispatch key, that pre_dispatch tracing now hooks into (cc @albanD @zou3519). I gave it a default implementation in C++ that no-ops and redispatches. But during pre_dispatch tracing with `make_fx(f, pre_dispatch=True)`, this PR now attaches `TorchProxyDispatchMode` directly to that key
- Having a dedicated key means that we don't have the "fallthrough problem" anymore, where we might push a mode onto a key that has a fallthrough kernel registered, so our mode never runs. So I removed all of the previous logic corresponding to making `TorchProxyDispatchMode` re-entrant, and pushing it onto the python mode stack (which was kind of gross).

File: aten/src/ATen/core/PythonFallbackKernel.cpp [hunk 1]
```diff
@@ -107,6 +107,11 @@ void pythonTLSSnapshotFallback(const c10::OperatorHandle &op, c10::DispatchKeySe
   op.redispatchBoxed(dispatch_keys & c10::DispatchKeySet(c10::DispatchKeySet::FULL_AFTER, c10::DispatchKey::PythonTLSSnapshot), stack);
 }
 
+// The PreDispatch key gets a no-op fallback that just redispatches.
+// The main way this key is used is that we can register a mode to it from python (e.g. TorchProxyDispatchMode, for pre_dispatch tracing)
```
Comment (albanD) [inline aten/src/ATen/core/PythonFallbackKernel.cpp:111]: nit: can you add a sentence explaining why this is a redispatch and not a fallthrough?

**Review by albanD — APPROVED** [review]
Sounds good!
Just a slightly lengthier comment on the alternatives and why we chose this one would be great.

**Comment by zou3519** [comment]
> For now, I ended up sticking our proxy mode on the mode stack corresponding to PythonTLSSnapshot, because it was simple and it works. 

We should update the commit message (the PR body still mentions the above)

**Review by zou3519 — APPROVED** [review]

**Comment by izaitsevfb** [comment]
@pytorchbot revert -m "Breaks internal builds see D46629736 TypeError: wrap_key() got an unexpected keyword argument pre_autograd" -c ghfirst

```
TypeError: wrap_key() got an unexpected keyword argument 'pre_autograd'
```

**Pull request closed without merging**